app/launcher.ts

```diff
@@ -0,0 +1,206 @@
+    console.warn({ e })
+    return false
+  }
+}
+
+interface IAction {
+  type: string
+  payload: unknown
+}
+
+export interface Message {
+  type: "message" | "error"
+  message: unknown
+}
+
+export class SiteLauncher {
+  public status: Status = GlobalStatus.NotStarted
+  public logs: Array<LogObject> = []
+  public rawLogs: Array<string> = []
+
+  private proc: ChildProcess | undefined
+  private listener?: (event: Message) => void
+
+  private logAction = (action: IAction): void => {
+    if (action.type === StructuredEventType.Log) {
+      this.logs.push(action.payload as LogObject)
+    }
+    if (action.type === StructuredEventType.SetStatus) {
+      console.log(`set status`, action.payload)
+      this.status = action.payload as Status
+    }
+    this.postMessage({
+      type: `message`,
+      message: { type: `LOG_ACTION`, action },
+    })
+  }
+
+  private handleExit = (code: number | null): void => {
+    this.logAction({ type: `EXIT`, payload: code || 0 })
+  }
+
+  private sendRawLog = (message: string): void => {
+    this.rawLogs.push(message)
+    this.logAction({ type: `RAW_LOG`, payload: message })
+  }
+
+  constructor(public root: string, public hash: string) {
+    // https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties
+  }
+
+  public setListener(listener: (event: Message) => void): void {
+    this.listener = listener
+  }
+
+  public removeListener(): void {
+    this.listener = undefined
+  }
+
+  private postMessage(event: Message): void {
+    this.listener?.(event)
+  }
+
+  private spawnProcess(
+    command: string,
+    args: Array<string> = [],
+    env: Record<string, string> = {}
+  ): ChildProcess {
+    const proc = fork(command, args, {
+      // The Gatsby process detects the IPC channel and uses it to send
+      // structured logs
+      stdio: [`pipe`, `pipe`, `pipe`, `ipc`],
+      cwd: this.root,
+      env: {
+        ...process.env,
+        FORCE_COLOR: `1`,
+        ...env,
+      },
+    })
+
+    proc.stderr?.setEncoding(`utf8`)
+    proc.stdout?.setEncoding(`utf8`)
+
+    proc.stderr?.on(`data`, this.sendRawLog)
+    proc.stdout?.on(`data`, this.sendRawLog)
+
+    proc.on(`message`, (message: IPCMessage) => {
+      if (message.type === IPCMessageType.LogAction) {
+        this.logAction(message.action as IAction)
+      }
+    })
+
+    return proc
+  }
+
+  public async start(clean: boolean = false): Promise<number> {
+    if (!(await isGatsbySite(this.root))) {
+      this.postMessage({
+        type: `error`,
+        message: `${this.root} is not a Gatsby site`,
+      })
+      console.log(`Not a gatsby site`)
+      return 0
+    }
+    console.log(`Is a gatsby site. Launching`)
+
+    if (this.proc) {
+      // We're restarting, so don't want to notify of exit
+      this.proc.off(`exit`, this.handleExit)
+
+      this.proc.kill()
+      this.proc = undefined
+    }
+
+    this.logs = []
+    this.rawLogs = []
+
+    const port = await detectPort(8000)
```

Comment:
what happens if the users project has a .env with a `PORT=3000` in it?

Reply:
I believe the command-line flag overrides the env var, so this one would "win".

Reply:
Yeah, I guess i'm wondering if this is problematic. Like do you think this will bother users if they built their app to develop on a different port?

Specifically if they are trying to expose port on 443 for SSL

Reply:
or 80 for ip exposing